### PR TITLE
feat(common): introduce StateTransferInitializerModule

### PR DESCRIPTION
**Pull request by Toxicable**
* **Please check if the PR fulfills these requirements**
```
- [x] The commit message follows
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
```

* **What modules are related to this pull-request**
```
- [x] common
```

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Adds a new module to help with StateTransfer

* **What is the current behavior?** (You can also link to an open issue here)
n/a

* **What is the new behavior (if this is a feature change)?**
Adds a new module to delay the app starting, therefore delay statetransfer accessing the dom till it's ready

### CODE REVIEW

File: modules/common/src/state-transfer-initilizer/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# State Transfer Initializer
+Delays the app bootstrap proces to ensure that the DOM content is loaded before
```
Comment (CaerusKaru) [inline modules/common/src/state-transfer-initilizer/README.md]: /s/proces/process

Also the thought trails off, maybe "before state transfer initialization"

File: modules/common/src/state-transfer-initilizer/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+Simply import the module into your project and you will not longer need to wrap your component bootsrap function in an `DOMContentLoaded` callback
```
Comment (CaerusKaru) [inline modules/common/src/state-transfer-initilizer/README.md]: /s/not/no

File: modules/common/src/state-transfer-initilizer/module.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { NgModule } from '@angular/core';
+
+export function domContentLoadedFactory(doc: Document) {
+  return () => {
```
Comment (CaerusKaru) [inline modules/common/src/state-transfer-initilizer/module.ts]: Can't this function take a Promise directly, ie cut out the inner return?

File: modules/common/src/state-transfer-initilizer/module.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    return new Promise ((resolve, reject) => {
+      const contentLoaded = () => {
+        resolve();
+        doc.removeEventListener('DOMContentLoaded', contentLoaded);
```
Comment (CaerusKaru) [inline modules/common/src/state-transfer-initilizer/module.ts]: Should this come before the resolve?

Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: yup

File: modules/common/src/state-transfer-initilizer/module.ts (hunk 3 above)
Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: Pretty sure it takes a fn which returns a Promise.
Will double check

Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: Here's an example https://github.com/angular/angular/blob/c8a1a14b87e5907458e8e87021e47f9796cb3257/packages/platform-webworker/src/web_workers/worker/location_providers.ts#L35

Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: That's what im doing

Comment (CaerusKaru) [reply modules/common/src/state-transfer-initilizer/module.ts]: I'm saying get rid of the first pair of curly braces

Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: I mean, it dosen't change anything but sure

File: modules/common/src/state-transfer-initilizer/module.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+      resolve();
+    };
+    doc.addEventListener('DOMContentLoaded', contentLoaded);
+  };
+}
+
+
+@NgModule({
+  providers: [
+    {provide: APP_INITIALIZER, multi: true, useFactory: domContentLoadedFactory, deps: [DOCUMENT]},
```
Comment (CaerusKaru) [inline modules/common/src/state-transfer-initilizer/module.ts]: Missed this:

Missing import for DOCUMENT

File: modules/common/src/state-transfer-initilizer/module.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  ]
+})
+export class StateTransferInitializerModule
```
Comment (CaerusKaru) [inline modules/common/src/state-transfer-initilizer/module.ts]: Also missing curly braces here

File: modules/common/src/state-transfer-initilizer/module.ts (hunk 5 above)
Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: :P

File: modules/common/src/state-transfer-initilizer/module.ts (hunk 6 above)
Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: needs testing yo

Comment (CaerusKaru) [reply modules/common/src/state-transfer-initilizer/module.ts]: I figured we'd add testing in a follow-up PR once we iron out our testing scheme

Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: yupyup that was my plan aswell

Comment (Toxicable) [reply modules/common/src/state-transfer-initilizer/module.ts]: I mean silly thiongs like that would be picked up there

File: modules/common/src/state-transfer-initilizer/module.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import { DOCUMENT } from '@angular/common';
+
+export function domContentLoadedFactory(doc: Document) {
+  return () => new Promise ((resolve, reject) => {
+    const contentLoaded = () => {
+      doc.removeEventListener('DOMContentLoaded', contentLoaded);
+      resolve();
+    };
+    doc.addEventListener('DOMContentLoaded', contentLoaded);
+  };
```
Comment (CaerusKaru) [inline modules/common/src/state-transfer-initilizer/module.ts]: Missing closing parens

**Pull request merged**